src/stan/callbacks/json_writer.hpp

```diff
@@ -0,0 +1,375 @@
+    }
+    // Shrink any unused space
+    new_value.resize(value.size() + count);
+    return new_value;
+  }
+
+  /**
+   * Writes key plus colon for key-value pair.
+   *
+   * @param[in] key member name.
+   */
+  void write_key(const std::string& key) { *output_ << "\"" << key << "\" : "; }
+  /**
+   * Writes a set of comma separated strings.
+   * Strings are cleaned to escape special characters.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  void write_vector(const std::vector<std::string>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      *output_ << process_string(*it) << ", ";
+    }
+    *output_ << v.back() << " ]";
+  }
+
+  /**
+   * Writes a set of comma separated values.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  template <class T>
+  void write_vector(const std::vector<T>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      *output_ << *it << ", ";
+    }
+    *output_ << v.back() << " ]";
+  }
+
+ public:
+  json_writer() : output_(nullptr) {}
+  /**
+   * Constructs a json writer with an output stream.
+   *
+   * @param[in, out] output unique pointer to a type inheriting from
+   * `std::ostream`
+   */
+  explicit json_writer(std::unique_ptr<Stream, Deleter>&& output)
+      : output_(std::move(output)) {
+    if (output_ == nullptr)
+      throw std::invalid_argument("writer cannot be null");
+  }
+
+  json_writer(json_writer& other) = delete;
+  json_writer(json_writer&& other) : output_(std::move(other.output_)) {}
+
+  ~json_writer() {}
+
+  /**
+   * Writes "{", initial token of a JSON record.
+   */
+  void begin_record() {
+    write_record_comma_if_needed();
+    *output_ << "{";
+    record_needs_comma_ = false;
+    record_depth_++;
+  }
+
+  /**
+   * Writes "\"key\" : {", initial token of a named JSON record.
+   * @param[in] key The name of the record.
+   */
+  void begin_record(const std::string& key) {
+    write_record_comma_if_needed();
+    *output_ << "\"" << key << "\": {";
+    record_needs_comma_ = false;
+    record_depth_++;
+  }
+  /**
+   * Writes "}", final token of a JSON record.
+   */
+  void end_record() {
+    if (record_depth_ > 0) {
+      *output_ << "}";
+      record_depth_--;
+      if (record_depth_ > 0) {
+        record_needs_comma_ = true;
+      }
+      record_internal_needs_comma_ = false;
+    } else {
+      throw std::runtime_error(
+          "Attempted to end record, but there is no open record.");
+    }
+  }
+
+  /**
+   * Writes "[", initial token of a JSON list.
+   */
+  void begin_list() {
+    *output_ << "[";
+    record_internal_needs_comma_ = false;
+  }
+  /**
+   * Writes "]", final token of a JSON list.
+   */
+  void end_list() {
+    *output_ << "]";
+    write_sep();
+    *output_ << "\n";
+  }
```

Comment:
Because all of the `write` functions assume a key is present, I think the only valid purpose for these functions is to construct a list of objects, which currently does not produce valid JSON.

Test case:
```c++
TEST_F(StanInterfaceCallbacksJsonWriter, list_of_objects) {
  std::string key("key");
  const int N = 5;
  std::vector<double> x;
  for (int n = 0; n < N; ++n)
    x.push_back(n);

  writer.begin_list();
  writer.begin_record();
  EXPECT_NO_THROW(writer.write(key, x));
  writer.end_record();

  writer.begin_record();
  EXPECT_NO_THROW(writer.write(key, x));
  writer.end_record();
  writer.end_list();

  EXPECT_EQ(
      "[{\"key\" : [ 0, 1, 2, 3, 4 ]}"
      ", {\"key\" : [ 0, 1, 2, 3, 4 ]}]",
      ss.str());

}
```

Currently, the comma is missing between the two objects.

Reply:
@SteveBronder and I also think we can probably just remove the begin/end list functions, as they're not strictly needed for our use cases. If we need them later we can add them then and do the extra heavy lifting to make them correct

Reply:
Def agree on removing begin and end list idt we use them anywhere in pathfinder

Reply:
gone